// @gilt-tech/node-client.svc-subscription/tests/support/app-fixture.js
'use strict';

module.exports = function() {
  var mockApp = {
    package: {
      gilt: {
        app: {
          principalGuid: 'b19955ff-17c8-4b76-91dc-17dbce5ab132'
        }
      }
    },
    log: {
      log: function () {

      }
    },
    utils: {
      request: function(url, opts) {

        if (/\/unsubscribe\/:opts.userGuid\/mailing_list\/:opts.mailingList\//.test(url)) {
          if (opts.userGuid === '37c14030-5364-4dd8-b84f-d9a2bf029498' && opts.mailingList === 'cart_reminder') {
            return Promise.resolve(JSON.parse('{"http":{"code":200,"message":"OK","headers":{"content-type":"application/json","date":"Wed, 17 May 2017 16:56:23 GMT","vary":"Accept-Encoding","x-gilt-service-id":"svc-subscription.5.ip-172-16-69-115","x-gilt-service-version":"svc-subscription-4.7.0","transfer-encoding":"chunked","connection":"keep-alive"},"httpVersion":"1.1"},"result":{"data":null,"error":null,"msg":"","nodeID":0,"nodename":"ip-172-16-69-115","request":"/subscription/1.0/unsubscribe/37c14030-5364-4dd8-b84f-d9a2bf029498/mailing_list/cart_reminder?principal_guid=d54acea6-f451-4d70-a48d-621aca330cb9","status":0,"timestamp":"Wed, 17 May 2017 16:56:23 UTC"}}'));
          } else {
            return Promise.resolve(JSON.parse('{"http":{"code":500,"message":"Internal Server Error","headers":{"content-type":"application/json","date":"Wed, 17 May 2017 16:52:24 GMT","vary":"Accept-Encoding","x-gilt-service-id":"svc-subscription.5.ip-172-16-68-131","x-gilt-service-version":"svc-subscription-4.7.0","transfer-encoding":"chunked","connection":"keep-alive"},"httpVersion":"1.1"},"result":{"data":{"id":"9988E27EE7844E23","message":"Internal Server Error"},"error":null,"msg":"","nodeID":0,"nodename":"ip-172-16-68-131","request":"/subscription/1.0/unsubscribe/37c14030-5364-4dd8-b84f-d9a2bf029498/mailing_list/cart_reminders?principal_guid=d54acea6-f451-4d70-a48d-621aca330cb9","status":0,"timestamp":"Wed, 17 May 2017 16:52:24 UTC"}}'));
          }
        }

        if (/\/unsubscribe_all\//.test(url)) {
          if (opts.userGuid === '37c14030-5364-4dd8-b84f-d9a2bf029498') {
            return Promise.resolve(JSON.parse('{"data":null,"error":null,"msg":"","nodeID":0,"nodename":"svc-subscription","request":"/subscription/1.0/unsubscribe_all/37c14030-5364-4dd8-b84f-d9a2bf029498?principal_guid=b19955ff-17c8-4b76-91dc-17dbce5ab132","status":0,"timestamp":"Tue, 16 May 2017 15:10:43 UTC"}'));
          } else {
            return Promise.resolve(JSON.parse('{"data":null,"error":{"errors":[{"code":0,"detail":"For input string: \"d9a2bf0294*8\"","field":"guid","index":null,"trigger":null}],"id":"CB375FE3E2D0ACE5","message":"invalid path param","trigger":null},"msg":"","nodeID":0,"nodename":"ip-172-16-68-131","request":"/subscription/1.0/unsubscribe_all/37c14030-5364-4dd8-b84f-d9a2bf0294*8?principal_guid=d54acea6-f451-4d70-a48d-621aca330cb9","status":1,"timestamp":"Tue, 16 May 2017 15:14:39 UTC"}'));
          }
        }

        if (/\/subscribe\//.test(url)) {
          if (opts.userGuid === '37c14030-5364-4dd8-b84f-d9a2bf029498' && opts.mailingList === 'cart_reminder') {
            return Promise.resolve(JSON.parse('{"data":null,"error":null,"msg":"","nodeID":0,"nodename":"ip-172-16-68-131","request":"/subscription/1.0/subscribe/37c14030-5364-4dd8-b84f-d9a2bf029498/mailing_list/cart_reminder?principal_guid=d54acea6-f451-4d70-a48d-621aca330cb9","status":0,"timestamp":"Tue, 16 May 2017 15:21:58 UTC"}'));
          } else {
            return Promise.resolve(JSON.parse('{"data":{"id":"1AB52403CC9671B9","message":"Internal Server Error"},"error":null,"msg":"","nodeID":0,"nodename":"ip-172-16-69-191","request":"/subscription/1.0/subscribe/37c14030-5364-4dd8-b84f-d9a2bf029498/mailing_list/cart_reminders?principal_guid=d54acea6-f451-4d70-a48d-621aca330cb9","status":0,"timestamp":"Tue, 16 May 2017 15:23:43 UTC"}'));
          }
        }
      },

      newrelic: {
        noticeError: function() {}
      }
    }
  };

  mockApp.log.info = mockApp.log.warn = mockApp.log.debug = mockApp.log.error = mockApp.log.log;

  return mockApp;
}();
